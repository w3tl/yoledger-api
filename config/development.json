{
  "credentials": {
    "algorithm": "dev"
  },
  "token": {
    "secret": "jwtsecret",
    "expiresIn": 3000
  }
}
