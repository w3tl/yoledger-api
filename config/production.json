{
  "credentials": {
    "algorithm": "argon2"
  },
  "port": "3030"
}
