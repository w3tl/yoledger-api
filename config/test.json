{
  "mongoUri": "mongodb://localhost:31000/?replicaSet=rs",
  "credentials": {
    "algorithm": "dev"
  },
  "token": {
    "secret": "jwtsecret",
    "expiresIn": 5
  }
}
