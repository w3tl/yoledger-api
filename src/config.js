const convict = require('convict');

const config = convict({
  env: {
    doc: 'The application environment.',
    format: ['production', 'development', 'test', 'stage'],
    default: 'development',
    env: 'NODE_ENV',
  },
  port: {
    doc: 'The port to bind.',
    format: 'port',
    default: 3030,
    env: 'GRAPHQL_PORT',
    arg: 'port',
  },
  path: {
    doc: 'Custom path',
    env: 'GRAPHQL_PATH',
    default: 'graphql',
  },
  credentials: {
    algorithm: {
      doc: 'Current crypt algorithm',
      env: 'CRED_ALG',
      default: 'default',
    },
  },
  token: {
    secret: {
      doc: 'Secret used for jwt tokens',
      format: '*',
      default: '',
      sensitive: true,
      env: 'JWT_SECRET',
    },
    expiresIn: {
      doc: 'Expressed in seconds describing a time span',
      format: 'int',
      default: 60, // 1m
      env: 'TOKEN_EXPIRES',
    },
  },
  mongoUri: {
    doc: 'MongoDB connection string',
    format: '*',
    default: 'mongodb://localhost',
    env: 'MONGO_URI',
  },
});

// Load environment dependent configuration
const env = config.get('env');
config.loadFile(`config/${env}.json`);
// Perform validation
config.validate({ allowed: 'strict' });

module.exports = config;
