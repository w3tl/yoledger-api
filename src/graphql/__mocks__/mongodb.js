import {
  users,
  transactions as transactionsJson,
  accounts,
  budgets as budgetsJson,
} from '../mocks';

const mongodb = jest.genMockFromModule('mongodb');

const addCreatedAt = ({ createdAt, ...other }) => ({ ...other, createdAt: new Date(createdAt) });
const addDate = ({ date, ...other }) => ({ ...other, date: new Date(date) });
const transactions = transactionsJson.map(addCreatedAt);
const budgets = budgetsJson.map(addDate);

const mockSession = {
  startTransaction: () => Promise.resolve(),
  commitTransaction: () => Promise.resolve(),
  abortTransaction: () => Promise.reject({ // eslint-disable-line prefer-promise-reject-errors
    errorLabels: ['TransientTransactionError'],
  }),
};

const mockCollections = {
  users: {
    findOne: ({ _id }) => Promise.resolve(users.find(u => u._id === _id)),
    updateOne: () => Promise.resolve({ result: { nModified: 1 } }),
  },
  transactions: {
    find: () => ({
      skip: skip => ({
        limit: limit => ({
          toArray: () => transactions.slice(skip, skip + limit),
        }),
      }),
    }),
    findOne: ({ _id }) => Promise.resolve(transactions.find(t => t._id === _id)),
    insertOne: (transaction) => {
      transactions.push({ _id: '50', ...transaction });
      return Promise.resolve({ insertedId: '50' });
    },
    findOneAndDelete: ({ id }) => Promise.resolve({ value: transactions.find(t => t.id === id) }),
    updateOne: () => Promise.resolve(),
  },
  accounts: {
    find: query => ({
      toArray: () => {
        if ('type' in query) {
          return Promise.resolve(accounts.filter(acc => query.type === acc.type));
        }
        if ('name' in query) {
          if ('$in' in query.name) {
            const result = accounts.filter(({ name }) => query.name.$in.indexOf(name) !== -1);
            return Promise.resolve(result);
          }
        }
        return Promise.resolve([]);
      },
    }),
    findOne: ({ name, _id }) => Promise.resolve(
      accounts.find(acc => name === acc.name || _id === acc._id),
    ),
    insertOne: (account) => {
      accounts.push({ _id: '44', ...account });
      return Promise.resolve({ insertedId: '44' });
    },
    updateOne: () => Promise.resolve(),
  },
  budget: {
    find: ({ $and, date, ...other }) => ({
      toArray: () => {
        if ($and) {
          return Promise.resolve(
            budgets.filter(b => (
              b.account.name === other['account.name']
              && b.date.getTime() >= $and[0].date.$gte.getTime()
              && b.date.getTime() <= $and[1].date.$lte.getTime()
            )),
          );
        }
        return Promise.resolve(budgets.filter(b => b.date.getTime() === new Date(date).getTime()));
      },
    }),
    findOne: ({ account, date }) => {
      if (account) { return Promise.resolve(budgets.find(t => t.account.name === account)); }
      if (date) { return Promise.resolve(budgets.find(t => t.date.getTime() === date.getTime())); }
      return budgets[0];
    },
    updateOne: ({ date, account }, { $set }) => {
      const idx = budgets
        .findIndex(b => b.date.getTime() === date.getTime() && b.account.name === account.name);
      if (idx) {
        budgets[idx] = { ...budgets[idx], ...$set.modificator };
      }
      return Promise.resolve({ upsertedId: budgets[idx]._id });
    },
    aggregate: () => ({
      toArray: () => [
        accounts.find(a => a.name === 'Food'),
        accounts.find(a => a.name === 'Train'),
      ],
    }),
  },
};


mongodb.connection = {
  db: () => ({
    collection: name => mockCollections[name],
  }),
  withSession: callback => callback(mockSession),
};

mongodb.ObjectId = id => String(id);

module.exports = mongodb;
