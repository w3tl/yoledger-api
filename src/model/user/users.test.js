jest.mock('./auth');
const auth = require('./auth');

auth.createCredentials.mockReturnValue({
  algorithm: 'very strong algorithm',
  createdAt: new Date('2018-06-20'),
  salt: 'random last',
  hash: 'crypto hash',
  valid: true,
});

import { MongoClient } from 'mongodb';
import config from '../../config';
import User from './index';

const RealDate = Date;
function mockDate(isoDate) {
  global.Date = class extends RealDate {
    constructor() {
      super();
      return new RealDate(isoDate);
    }
  };
}

let connection;
let db;

describe('user', () => {
  beforeAll(async () => {
    mockDate('2018-06-20');
    connection = await MongoClient.connect(config.get('mongoUri'), {
      connectWithNoPrimary: true,
      useNewUrlParser: true,
    });
    db = connection.db('dbusers');
    const userModel = new User(db, 'dumb');
    await userModel.init();
  });

  afterAll(async () => {
    global.Date = RealDate;
    await connection.close();
  });

  test('create new user', async () => {
    const userModel = new User(db, 'admin');
    const user = await userModel.create({ password: 'password' });
    expect(user).toMatchSnapshot();
  });

  describe('validatePassword', () => {
    beforeAll(async () => {
      const userModel = new User(db, 'dumb');
      await userModel.clear();
      await userModel.create({ password: 'password' });
    });

    test('should throw error when user not found', async () => {
      const userModel = new User(db, 'wronUser');
      expect(userModel.validatePassword('wronUser'))
        .rejects
        .toThrowError(/User not found/);
    });

    test('should return false when validateCreadentials throw error', async () => {
      auth.validateCreadentials.mockImplementation(() => {
        throw new Error();
      });
      const userModel = new User(db, 'dumb');
      const result = await userModel.validatePassword('dumb', 'passwd');
      expect(result).toBeFalsy();
    });

    test('should return true when password valid', async () => {
      auth.validateCreadentials.mockReturnValue(true);
      const userModel = new User(db, 'dumb');
      expect(userModel.validatePassword('dumb', 'password')).toBeTruthy();
    });
  });

  describe('changePassword', () => {
    it('should return false when password already used', async () => {
      const userModel = new User(db, 'dumb1');
      await userModel.insertOne({
        _id: 'dumb1',
        credentials: {
          current: {
            algorithm: 'very strong algorithm',
            createdAt: new Date('2018-06-20'),
            salt: 'random salt',
            hash: 'hash to archive',
            valid: true,
          },
          store: [{
            algorithm: 'very strong algorithm',
            createdAt: new Date('2018-06-20'),
            salt: 'random salt',
            hash: 'old crypto hash',
            valid: true,
          }],
        },
        createdAt: new Date('2018-06-20'),
      });
      auth.validateCreadentials.mockReturnValue(Promise.resolve(true));
      const result = await userModel.changePassword('password');
      expect(result).toBeFalsy();
    });

    it('should return true when password are new and update cred\'s store', async () => {
      const userModel = new User(db, 'dumb2');
      await userModel.insertOne({
        _id: 'dumb2',
        credentials: {
          current: {
            algorithm: 'very strong algorithm',
            createdAt: new Date('2018-06-20'),
            salt: 'random salt',
            hash: 'hash to archive',
            valid: true,
          },
          store: [{
            algorithm: 'very strong algorithm',
            createdAt: new Date('2018-06-20'),
            salt: 'random salt',
            hash: 'old crypto hash',
            valid: true,
          }],
        },
        createdAt: new Date('2018-06-20'),
      });
      auth.validateCreadentials.mockReturnValue(Promise.resolve(false));
      const result = await userModel.changePassword('password');
      expect(result).toBeTruthy();
      const user = await userModel.findOne();
      expect(user).toMatchSnapshot();
    });
  });
});
