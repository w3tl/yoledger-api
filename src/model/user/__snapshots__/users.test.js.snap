// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user changePassword should return true when password are new and update cred's store 1`] = `
Object {
  "_id": "dumb2",
  "createdAt": 2018-06-20T00:00:00.000Z,
  "credentials": Object {
    "current": Object {
      "algorithm": "very strong algorithm",
      "createdAt": 2018-06-20T00:00:00.000Z,
      "hash": "crypto hash",
      "salt": "random last",
      "valid": true,
    },
    "store": Array [
      Object {
        "algorithm": "very strong algorithm",
        "createdAt": 2018-06-20T00:00:00.000Z,
        "hash": "old crypto hash",
        "salt": "random salt",
        "valid": true,
      },
      Object {
        "algorithm": "very strong algorithm",
        "createdAt": 2018-06-20T00:00:00.000Z,
        "hash": "hash to archive",
        "salt": "random salt",
        "valid": true,
      },
    ],
  },
  "userId": "dumb2",
}
`;

exports[`user create new user 1`] = `
Object {
  "_id": "admin",
  "createdAt": 2018-06-20T00:00:00.000Z,
  "credentials": Object {
    "current": Object {
      "algorithm": "very strong algorithm",
      "createdAt": 2018-06-20T00:00:00.000Z,
      "hash": "crypto hash",
      "salt": "random last",
      "valid": true,
    },
    "store": Array [],
  },
  "userId": "admin",
}
`;
